### PR TITLE
[FEAT][executors]: implement stateless zero-KV scoring executor

### PR DESCRIPTION
**Pull request by inaniloquentee**
## Summary
- Add `StatelessForwardExecutor` for no-cache reference/reward scoring.
- Add a local paged-KV scoring baseline and smoke benchmark for footprint comparison.
- Wire stateless rewards/reference logprobs into rollout payloads and trainer batch construction.
- Add unit and HF integration coverage for the new scoring path.

## Validation
- DCO: commit includes `Signed-off-by: inaniloquentee <3051000145@qq.com>`
- `py -3.13 -m pre_commit run --all-files`
- `py -3.13 -m mypy --ignore-missing-imports rl_engine/`
- `py -3.13 -m pytest rl_engine/tests/test_dispatch.py -v`
- `py -3.13 -m mkdocs build --strict -f mkdocs.yaml`
- `py -3.13 -m pytest -q` (`174 passed, 1 skipped`)
- `py -3.13 benchmarks\benchmark_stateless_executor.py --smoke --mode both --compare-paged-kv`

## Notes
- Real vLLM runtime is not available in the local `cp313/win_amd64` environment, so the benchmark includes a reproducible local paged-KV comparison path.

## Summary by CodeRabbit

* **New Features**
  * Added a stateless scoring executor supporting reference log-probabilities, rewards, and combined scoring.
  * Added a paged KV reservation scoring baseline with reservation/cache metrics.
  * Added a CLI benchmark tool for stateless scoring with optional paged-KV comparison.
  * Integrated stateless scoring into the training pipeline with grouped candidates and payload-backed reward/log-prob handling.
* **Bug Fixes**
  * Improved KL reference log-prob computation and advantage metrics to reflect only active completion tokens.
* **Tests**
  * Added unit/integration coverage for stateless scoring, paged KV reservation, and heavy-runtime import isolation.

### CODE REVIEW

**Comment by Flink-ddd** [comment]
Excellent work! Thank you for your contribution, Could you add an issue to this PR? If this issue exists, please do so. otherwise, please create one.

**Comment by inaniloquentee** [comment]
> Excellent work! Thank you for your contribution, Could you add an issue to this PR? If this issue exists, please do so. otherwise, please create one.

This PR already has a corresponding issue: https://github.com/RL-Align/RL-Kernel/issues/47

File: rl_engine/executors/stateless_executor.py [hunk 1]
```diff
@@ -0,0 +1,564 @@
...
+        temperature=temperature,
+        output_dtype=output_dtype,
+    )
+    result = torch.zeros(
+        inputs.input_ids.shape,
+        device=logits.device,
+        dtype=output_dtype,
+    )
+    result[:, 1:] = shifted_logps
+    return result.masked_fill(~_bool_mask(inputs.completion_mask, device=result.device), 0.0)
```
Comment (Flink-ddd) [inline rl_engine/executors/stateless_executor.py:253]: In score_reference_logprobs, result[:, 0] is always 0 but the un-shifted completion_mask is applied afterward, so any sequence with completion_mask[:, 0] == True gets its logprob silently zeroed. This is an undocumented precondition. Suggest adding to _validate_inputs:

```
if bool(_bool_mask(inputs.completion_mask, device=input_ids.device)[:, 0].any().item()):
    raise ValueError("completion_mask[:, 0] must be False; completions cannot start at position 0")
```

File: rl_engine/executors/stateless_executor.py [hunk 2]
```diff
@@ -0,0 +1,564 @@
...
+
+    use_cache_targets = 0
+    attention_targets = 0
+
+    for target in _model_config_targets(model):
+        if hasattr(target, "use_cache"):
+            target.use_cache = False
+            use_cache_targets += 1
+        if config.attention_backend != "model_default":
+            target.attn_implementation = config.attention_backend
```
Comment (Flink-ddd) [inline rl_engine/executors/stateless_executor.py:355]: configure_stateless_model mutates model.config (and generation_config) in place permanently — if the same model instance is reused elsewhere its attention backend / use_cache stay overwritten with no restore. Either document this side effect, or wrap score() in a restore-on-exit context manager:

```
from contextlib import contextmanager
_MISSING = object()

@contextmanager
def _temporarily_configure(model, config):
    saved = [
        (t, getattr(t, "use_cache", _MISSING), getattr(t, "_attn_implementation", _MISSING))
        for t in _model_config_targets(model)
    ]
    try:
        configure_stateless_model(model, config)
        yield
    finally:
        for t, uc, attn in saved:
            if uc is not _MISSING:
                t.use_cache = uc
            if attn is not _MISSING:
                t._attn_implementation = attn
```

File: rl_engine/executors/stateless_executor.py [hunk 3]
```diff
@@ -0,0 +1,564 @@
...
+        return False
+    if isinstance(exc, (ImportError, ModuleNotFoundError, StopIteration)):
+        return True
+    message = str(exc).lower()
+    markers = (
+        "flash_attention",
+        "flash attention",
+        "flash-attn",
+        "flash_attn",
+        "kernels",
```
Comment (Flink-ddd) [inline rl_engine/executors/stateless_executor.py]: In _should_fallback_attention_backend, the marker "kernels" is too broad, it will match unrelated errors (e.g. a CUDA kernel launch failure) and silently fall back to eager, hiding the real exception. Remove "kernels" or tighten it to something like "flash attention kernels".

File: rl_engine/executors/deepspeed_trainer.py [hunk 4]
```diff
@@ -145,6 +147,16 @@ def train(self, rollout: RolloutStageResult) -> TrainingStageResult:
                 "training_device": str(self.device),
                 "deepspeed_engine": type(self.engine).__name__,
                 "deepspeed_zero_stage": self.config.zero_stage,
+                "advantage_mean": (
+                    float(active_advantages.mean().detach().cpu().item())
+                    if active_advantages.numel()
+                    else 0.0
+                ),
+                "advantage_std": (
```
Comment (Flink-ddd) [inline rl_engine/executors/deepspeed_trainer.py]: Advantages are already group-normalized in _compute_group_relative_advantages, so taking mean/std again over all flattened active tokens is neither the per-group std nor a standard GRPO statistic. unbiased=False is fine, but rename to make scope explicit, e.g. active_advantage_std_global, to avoid misreading.

File: benchmarks/benchmark_stateless_executor.py [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+            f"zero_kv_cache={result.metrics['zero_kv_cache']};"
+            f"attention_backend={result.metrics['attention_backend']};"
+            f"hidden_dim={args.hidden_dim};vocab_size={args.vocab_size};{_environment()}"
+        ),
+    }
+    if args.compare_paged_kv:
+        paged_result = _score_paged_kv(args, inputs, model)
+        paged_metrics = paged_result.metrics
+        row["paged_kv_baseline_mb"] = f"{float(paged_metrics['paged_kv_cache_reserved_mb']):.4f}"
+        savings_mb = float(paged_metrics["total_kv_cache_mb"]) - float(
```
Comment (Flink-ddd) [inline benchmarks/benchmark_stateless_executor.py:266]: The zero_kv_cache_savings_mb value in _run measures savings vs. a synthetic local reservation, not real vLLM pre-allocation (the PR description acknowledges this). Suggest renaming the column in CSV_COLUMNS to savings_vs_local_reservation_baseline so it isn't reported as real vLLM savings.

**Review by Flink-ddd — CHANGES_REQUESTED** [review]
Good approach and test coverage, happy to re-review once fixed.

**Comment by inaniloquentee** [comment]
> Good approach and test coverage, happy to re-review once fixed.

Thanks for the careful review! I’ve addressed the requested changes and added/updated coverage for them. The relevant tests are passing now.

**Review by Flink-ddd — APPROVED** [review]
LGTM now, Thank you for your update.

**Review by bitborne — APPROVED** [review]
LGTM！

**Comment by frank-2077** [comment]
LGTM！

**Review by frank-2077 — APPROVED** [review]
